BioFVM/BioFVM_microenvironment.cpp

```diff
@@ -473,212 +473,237 @@ void Microenvironment::resize_densities( int new_size )
-	default_microenvironment_options.Dirichlet_xmax.push_back( false ); 
-	default_microenvironment_options.Dirichlet_ymin.push_back( false ); 
-	default_microenvironment_options.Dirichlet_ymax.push_back( false ); 
-	default_microenvironment_options.Dirichlet_zmin.push_back( false ); 
-	default_microenvironment_options.Dirichlet_zmax.push_back( false ); 
-	
-	default_microenvironment_options.Dirichlet_xmin_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_xmax_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_ymin_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_ymax_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_zmin_values.push_back( 1.0 ); 
-	default_microenvironment_options.Dirichlet_zmax_values.push_back( 1.0 ); 	
+		// resize the gradient data structures
+		for( unsigned int k=0 ; k < mesh.voxels.size() ; k++ )
+		{
+			gradient_vectors[k].resize( number_of_densities() );
+			for( unsigned int i=0 ; i < number_of_densities() ; i++ )
+			{
+				(gradient_vectors[k])[i].resize( 3, 0.0 );
+			}
+		}
+		gradient_vector_computed.resize( mesh.voxels.size() , false );
+
+		one_half = one;
+		one_half *= 0.5;
+
+		one_third = one;
+		one_third /= 3.0;
+
+		dirichlet_value_vectors.assign( mesh.voxels.size(), one );
+		dirichlet_activation_vector.push_back( false );
+		dirichlet_activation_vectors.assign( mesh.voxels.size(), dirichlet_activation_vector );
+
+		// fix in PhysiCell preview November 2017
+		default_microenvironment_options.Dirichlet_condition_vector.push_back( 1.0 ); // = one;
+		default_microenvironment_options.Dirichlet_activation_vector.push_back( false ); // assign( number_of_densities(), false );
+
+		default_microenvironment_options.initial_condition_vector.push_back( 1.0 );
+
+		default_microenvironment_options.Dirichlet_all.push_back( true );
+//		default_microenvironment_options.Dirichlet_interior.push_back( true );
+		default_microenvironment_options.Dirichlet_xmin.push_back( false );
+		default_microenvironment_options.Dirichlet_xmax.push_back( false );
+		default_microenvironment_options.Dirichlet_ymin.push_back( false );
+		default_microenvironment_options.Dirichlet_ymax.push_back( false );
+		default_microenvironment_options.Dirichlet_zmin.push_back( false );
+		default_microenvironment_options.Dirichlet_zmax.push_back( false );
+
+		default_microenvironment_options.Dirichlet_xmin_values.push_back( 1.0 );
+		default_microenvironment_options.Dirichlet_xmax_values.push_back( 1.0 );
+		default_microenvironment_options.Dirichlet_ymin_values.push_back( 1.0 );
+		default_microenvironment_options.Dirichlet_ymax_values.push_back( 1.0 );
+		default_microenvironment_options.Dirichlet_zmin_values.push_back( 1.0 );
+		default_microenvironment_options.Dirichlet_zmax_values.push_back( 1.0 );
+        }
+        else
+        {
+		std::cout << "Warning: density " << name << " alredy exist!" << std::endl;
```

Comment:
1. I think we ought to throw an error here. I know that's not strictly backwards compatible , but we should force the user to reckon with adding two identical substrates.
2. For readability, let's have the current `else` block be the first case (`if (find_density_index( name ) >= 0)`) so that we can throw an error (or `return` if that's what we choose) and then do away with the brackets afterwards